### DISCUSSION TITLE
About ShutDown

### DISCUSSION THREAD

**Discussion by cloudandzack007** [OpenDDS/OpenDDS#3731] [Q&A] [answered]
Hello
I create a dll with cpp to use opendds.
The dll include a interface [Release]

```C++
void Release()
{
	m_participant->delete_contained_entities();
	m_dpf->delete_participant(m_participant);
	TheServiceParticipant->shutdown();
}
```

If I use C# to call it (WPF or U3D), it will crush.
If I use Cpp to call it (Qt app), it will be normal.
Why is this so?

**Comment by iguessthislldo** [comment, marked as answer]
I think more info is probably needed. Can you post the stack trace from the crash?

**Reply by cloudandzack007** [reply]
I find the reason and I think i find a bug.
In the fun  [void DataReaderListener::on_subscription_matched(
	DDS::DataReader_ptr reader,
	const DDS::SubscriptionMatchedStatus& status)]
I use the [reader->get_matched_publication_data] to get data (I find DDS::BuiltinTopicKey_t key include mac_address, this is my app want to get).
When [m_participant->delete_contained_entities();] is called, it wiill trigger [on_subscription_matched] and trigger [reader->get_matched_publication_data].
Then the OpenDDS will run [DDS::ReturnCode_t instance_handle_to_bit_data(
  DomainParticipantImpl* dp,
  const char* bit_name,
  const DDS::InstanceHandle_t& handle,
  BIT_DataSeq& data)]
and run [DDS::DataReader_var reader = bit_subscriber->lookup_datareader(bit_name);].
The bit_name is "DCPSPublication".
But, the return value [reader] is null_ptr (Because the fun [delete_contained_entities()] delete it)
The code don't check it and continue and the program crash.
I add the code 
[if (bit_reader == nullptr) return DDS::RETCODE_BAD_PARAMETER;]
to fix this.
But I think you can fix it by more scientific methods.

**Reply by iguessthislldo** [reply]
Yes, I think that makes sense. It might be possible to us avoid getting in that situation, but I added the check you suggested in #3741 so at least it doesn't crash outright.
